src/browser-installer/registry/index.ts

```diff
@@ -0,0 +1,215 @@
+import type { BrowserPlatform } from "@puppeteer/browsers";
+import { readJsonSync, outputJSONSync, existsSync } from "fs-extra";
+import path from "path";
+import {
+    getRegistryPath,
+    browserInstallerDebug,
+    Driver,
+    Browser,
+    getMilestone,
+    normalizeChromeVersion,
+    semverVersionsComparator,
+    type SupportedBrowser,
+    type SupportedDriver,
+    type DownloadProgressCallback,
+} from "../utils";
+import { getFirefoxBuildId } from "../firefox/utils";
+import logger from "../../utils/logger";
+import type { createBrowserDownloadProgressBar } from "./cli-progress-bar";
+
+type VersionToPathMap = Record<string, string | Promise<string>>;
+type BinaryName = Exclude<SupportedBrowser | SupportedDriver, SupportedBrowser & SupportedDriver>;
+type RegistryKey = `${BinaryName}_${BrowserPlatform}`;
+type Registry = Record<RegistryKey, VersionToPathMap>;
+
+const registryPath = getRegistryPath();
+const registry: Registry = existsSync(registryPath) ? readJsonSync(registryPath) : {};
```

Comment:
Registry is:
- Singletone
- Sync

So its state is valid and up-to-date all the time.
It acts like synchronization point  between multiple browsers. So when multiple browsers (possibly with different browserId) with same browserName and browserVersion would want to install the browser, `install` function only being called one time, and other browsers (NewBrowser instance) receive same promise, which will be resolved to binary path when download is complete

Reply:
It looks like this:
```json
{
    "chrome_mac_arm": {
        "114.0.5735.133": "../browsers/chrome/mac_arm-114.0.5735.133/chrome-mac-arm64/Google Chrome for Testing.app/Contents/MacOS/Google Chrome for Testing",
        "115.0.5790.170": "../browsers/chrome/mac_arm-115.0.5790.170/chrome-mac-arm64/Google Chrome for Testing.app/Contents/MacOS/Google Chrome for Testing"
    },
    "chromedriver_mac_arm": {
        "110.0.5481.77": "../drivers/chromedriver/mac_arm-110.0.5481.77/chromedriver",
        "114.0.5735.90": "../drivers/chromedriver/mac_arm-114.0.5735.90/chromedriver",
        "115.0.5790.170": "../drivers/chromedriver/mac_arm-115.0.5790.170/chromedriver-mac-arm64/chromedriver",
        "92.0.4515.107": "../drivers/chromedriver/mac_arm-92.0.4515.107/chromedriver"
    },
    "chromium_mac": {
        "92": "../browsers/chromium/mac-885361/chrome-mac/Chromium.app/Contents/MacOS/Chromium"
    },
    "chromium_mac_arm": {
        "110": "../browsers/chromium/mac_arm-1084160/chrome-mac/Chromium.app/Contents/MacOS/Chromium"
    },
    "firefox_mac_arm": {
        "stable_115.0": "../browsers/firefox/mac_arm-stable_115.0/Firefox.app/Contents/MacOS/firefox"
    },
    "geckodriver_mac_arm": {
        "0.35.0": "../drivers/geckodriver/mac_arm-0.35.0/geckodriver"
    }
}
```
Here we install "mac" version for chromium@92 and "mac_arm" for chromium@110, because there is no "mac_arm" version for chromium@92, even when there is chromedriver@92 for "mac_arm"